lib/rpc/src/js_tracer/tracer.rs

```diff
@@ -0,0 +1,835 @@
+            self.current_depth -= 1;
+        }
+
+        let obj = serde_json::json!({
+            "gasUsed": gas_used,
+            "output": output.map(|o| o.encode_hex()),
+            "error": revert_reason
+        });
+        self.invoke_method(TracerMethod::Exit, &obj);
+    }
+
+    fn on_storage_read(&mut self, _: bool, _: Address, _: B256, _: B256) {}
+
+    fn on_storage_write(&mut self, _is_transient: bool, address: Address, key: B256, value: B256) {
+        {
+            let mut overlay = self.storage_overlay.borrow_mut();
+            match overlay.entry((address, key)) {
+                Entry::Occupied(mut entry) => {
+                    let slot = entry.get_mut();
+                    slot.previous = Some(slot.value);
+                    slot.value = value;
+                    slot.committed = false;
+                }
+                Entry::Vacant(vacant) => {
+                    vacant.insert(OverlayEntry::new_pending(value));
+                }
+            }
+        }
+        let obj = serde_json::json!({
+            "address": address,
+            "key": key,
+            "value": value,
+        });
+
+        // this method is an extension beyond geth tracer interface, convenient for state change tracking
+        self.invoke_method(TracerMethod::Write, &obj);
+    }
+
+    fn on_bytecode_change(
+        &mut self,
+        address: Address,
+        new_raw_bytecode: Option<&[u8]>,
+        _new_internal_bytecode_hash: B256,
+        new_observable_bytecode_length: u32,
+    ) {
+        let new_value = new_raw_bytecode.map(|code| {
+            let len = new_observable_bytecode_length as usize;
+            let slice = if code.len() >= len {
+                &code[..len]
+            } else {
+                code
+            };
+            slice.to_vec()
+        });
+
+        {
+            let mut overlay = self.code_overlay.borrow_mut();
+            match overlay.entry(address) {
+                Entry::Occupied(mut entry) => {
+                    let record = entry.get_mut();
+                    if record.committed && record.previous.is_none() {
+                        record.previous = Some(record.value.clone());
+                    }
+                    record.value = new_value.clone();
+                    record.committed = false;
+                }
+                Entry::Vacant(vacant) => {
+                    vacant.insert(OverlayEntry::new_pending(new_value));
+                }
+            }
+        }
+    }
+
+    fn on_event(&mut self, _: Address, _: Vec<B256>, _: &[u8]) {}
+
+    fn begin_tx(&mut self, _calldata: &[u8]) {
+        self.tx_failed = false;
+        self.current_depth = 0;
+        self.pending_step = None;
+        self.pending_create_type = None;
+        self.last_finished_frame = None;
+
+        let config = self.tracer_config.clone();
+        self.invoke_method(TracerMethod::Setup, &config);
+    }
+
+    fn finish_tx(&mut self) {
```

Comment:
We also should cleanup accounts marked for selfdestruct at this step. 
Note: in EVM we delete accounts only if SELFDESTRUCT is called in the same tx where the corresponding contract is deployed. So we do not destruct contracts that were deployed in a different tx